BaseTools/Plugin/RustEnvironmentCheck/RustEnvironmentCheck.py

```diff
@@ -0,0 +1,173 @@
+##
+import logging
+import re
+from collections import namedtuple
+from edk2toolext.environment.plugintypes.uefi_build_plugin import IUefiBuildPlugin
+from edk2toolext.environment.uefi_build import UefiBuilder
+from edk2toollib.utility_functions import RunCmd
+from io import StringIO
+
+RustToolInfo = namedtuple("RustToolInfo", ["presence_cmd", "install_help"])
+RustToolChainInfo = namedtuple("RustToolChainInfo", ["error", "toolchain"])
+
+
+class RustEnvironmentCheck(IUefiBuildPlugin):
+    """Checks that the system environment is ready to build Rust code."""
+
+    def do_pre_build(self, _: UefiBuilder) -> int:
+        """Rust environment checks during pre-build.
+
+        Args:
+            builder (UefiBuilder): A UEFI builder object for this build.
+
+        Returns:
+            int: The number of environment issues found. Zero indicates no
+            action is needed.
+        """
+        def verify_cmd(name: str, params: str = "--version") -> bool:
+            """Indicates if a command can successfully be executed.
+
+            Args:
+                name (str): Tool name.
+                params (str, optional): Tool params. Defaults to "--version".
+
+            Returns:
+                bool: True on success. False on failure to run the command.
+            """
+            cmd_output = StringIO()
+            ret = RunCmd(name, params, outstream=cmd_output,
+                         logging_level=logging.DEBUG)
+            return ret == 0
+
+        def verify_workspace_rust_toolchain_is_installed() -> RustToolChainInfo:
+            """Verifies the rust toolchain used in the workspace is available.
+
+            Note: This function does not use the toml library to parse the toml
+            file since the file is very simple and its not desirable to add the
+            toml module as a dependency.
+
+            Returns:
+                RustToolChainInfo: A tuple that indicates if the toolchain is
+                available and any the toolchain version if found.
+            """
+            WORKSPACE_TOOLCHAIN_FILE = "rust-toolchain.toml"
+
+            toolchain_version = None
+            try:
+                with open(WORKSPACE_TOOLCHAIN_FILE, 'r') as toml_file:
+                    content = toml_file.read()
+                    match = re.search(r'channel\s*=\s*"([^"]+)"', content)
+                    if match:
+                        toolchain_version = match.group(1)
+            except FileNotFoundError:
+                # If a file is not found. Do not check any further.
+                return RustToolChainInfo(error=False, toolchain=None)
+
+            if not toolchain_version:
+                # If the file is not in an expected format, let that be handled
+                # elsewhere and do not look further.
+                return RustToolChainInfo(error=False, toolchain=None)
+
+            installed_toolchains = StringIO()
+            ret = RunCmd("rustup", "toolchain list",

```

Comment:
Under linux, if cargo is installed through the distribution repositories, rustup is not available.

Do we need to clarify that rust must be installed from the `https://sh.rustup.rs`?